src/python/pants/util/objects.py

```diff
@@ -147,7 +147,22 @@ def __ne__(self, other):
     # explicitly implemented, otherwise Python will raise "unhashable type". See
     # https://docs.python.org/3/reference/datamodel.html#object.__hash__.
     def __hash__(self):
-      return super(DataType, self).__hash__()
+      try:
+        return super(DataType, self).__hash__()
+      except TypeError:
+        # If any fields are unhashable, we want to be able to specify which ones in the error
+        # message, but we don't want to slow down the normal __hash__ code path, so we take the time
+        # to break it down by field if we know the __hash__ fails for some reason.
+        for field_name, value in self._asdict().items():
+          try:
+            hash(value)
+          except TypeError as e:
+            raise TypeError("For datatype object {} (type '{}'): in field '{}': {}"
+                            .format(self, type(self).__name__, field_name, e))
+          else:
```

Comment:
I think `else` here is to run code after the `try` code; I think you just want to remove the `else` branch here, and it will do what you want.

Reply:
Because all the `else` clause is doing is calling `raise`, yes, here you can delete both the `else` and the `raise` lines for the same behavior.

If you had a different logic in the `else` clause, it would be idiomatic to use. @illicitonion yes, the way `try-except-else-finally` evaluates is first try block; if except is triggered, then except followed by finally; if except is not triggered, then else followed by finally.

Reply:
This was actually supposed to be a `for`/`else` (significant indentation is significant), which I've converted into a bare `raise`.